File: app/src/main/java/com/ftang/tightpenny/model/SpendingEntryRepository.java
package com.ftang.tightpenny.model;

import org.apache.commons.lang.math.Range;
import org.joda.time.DateTime;
import org.roboguice.shaded.goole.common.collect.Lists;
import org.roboguice.shaded.goole.common.collect.Maps;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Map;

public class SpendingEntryRepository {

    public Map<Category, List<SpendingEntry>> findSpendingsForDate(int year, int month, int day) {

        Map<Category, List<SpendingEntry>> categorizedEntries = Maps.newHashMap();

        List<SpendingEntry> entries = SpendingEntry.find(SpendingEntry.class,
                "year = ? and month = ? and day = ?", "" + year, "" + month, "" + day);

        for (SpendingEntry entry : entries) {
            Category category = Category.valueOf(entry.category);
            List<SpendingEntry> catEntries = categorizedEntries.get(category);
            if (catEntries == null) {
                catEntries = Lists.newArrayList();
                categorizedEntries.put(category, catEntries);
            }
            catEntries.add(entry);
        }
        return categorizedEntries;
    }

    public List<AggregateSpendingEntry> getMonthSpendings(int year, int month) {
        List<Map<Category, List<SpendingEntry>>> monthsEntries = new ArrayList<>();
        for (int day = 1; day <= 31; day++) {
            monthsEntries.add(findSpendingsForDate(year, month, day));
        }
        List<AggregateSpendingEntry> l = new ArrayList<AggregateSpendingEntry>();
        l.add(new AggregateSpendingEntry("category", new ArrayList<BigDecimal>()));
        return l;
        //return monthsEntries.get(0); // FIXME
    }

    public void addNewSpending(Category category, BigDecimal amount) {
        DateTime now = DateTime.now();

        SpendingEntry entry = new SpendingEntry("" + category.getId(), amount, now.toDate().getTime(),
                now.getYear(), now.getMonthOfYear(), now.getDayOfMonth());
        entry.save();
    }
}

File: app/src/main/java/com/ftang/tightpenny/model/SpendingEntryRepository.kt
package com.ftang.tightpenny.model

import android.util.Log
import org.joda.time.DateTime

import java.math.BigDecimal
import java.util.*

class SpendingEntryRepository {

    val TAG = "TightPenny-DAO"

    fun clearDb(): Unit {
        SpendingEntry.deleteAll(SpendingEntry::class.java)
    }

    fun findSpendingsForDate(year: Int, month: Int, day: Int): Map<Category, List<SpendingEntry>> {

        val entries = SpendingEntry.find(SpendingEntry::class.java,
                "year = ? and month = ? and day = ?", "" + year, "" + month, "" + day)

        return entries.groupBy { it.category }.mapKeys { Category.valueOf(it.component1().toInt()) }
    }

    fun getMonthSpendings(year: Int, month: Int): List<AggregateSpendingEntry> {
        val monthsEntries = (1..31).map { day ->
            findSpendingsForDate(year, month, day)
        }

        val results = monthsEntries.flatMap {it.toList() }.groupBy { it.first }.mapValues {
            it.component2().flatMap { it.second }
        }.toList().map { AggregateSpendingEntry(it.first, it.second.map { SimpleSpendingEntry.fromSpendingEntry(it) }.toArrayList())}

        return results
    }

    fun addNewSpending(category: Category, amount: BigDecimal): SpendingEntry {
        val now = DateTime.now()

        val entry = SpendingEntry(category, amount, now.toDate().time,
                now.year, now.monthOfYear, now.dayOfMonth)
        entry.save()
        return entry
    }

    fun removeSpending(uuid: String) {
        SpendingEntry.find(SpendingEntry::class.java, "uuid = ?", uuid)
                .forEach {
                    Log.d(TAG, "Trying to delete " + it)
                    it.delete()
                }
    }
}
